packages/core/src/hydration/i18n.ts

```diff
@@ -0,0 +1,325 @@
+  }
+}
+
+/**
+ * Skip over some nodes during hydration.
+ *
+ * Note: we use this instead of `siblingAfter` as it's expected that
+ * sometimes we might encounter null nodes. In those cases, we want to
+ * defer to downstream error handling to provide proper context.
+ */
+function skipHydrationNodes(state: I18nHydrationState, skip: number) {
+  let currentNode = state.currentNode;
+  for (let i = 0; i < skip; i++) {
+    if (!currentNode) {
+      break;
+    }
+    currentNode = currentNode?.nextSibling ?? null;
+  }
+  state.currentNode = currentNode;
+}
+
+/**
+ * Fork the given state into a new state for hydrating children.
+ */
+function forkChildHydrationState(state: I18nHydrationState) {
+  const currentNode = state.currentNode as Node | null;
+  return {currentNode: currentNode?.firstChild ?? null, isConnected: state.isConnected};
+}
+
+function prepareI18nBlockForHydrationImpl(lView: LView, index: number) {
+  if (!isI18nHydrationSupportEnabled()) {
+    return;
+  }
+
+  const hydrationInfo = lView[HYDRATION];
+  if (!hydrationInfo) {
+    return;
+  }
+
+  const tView = lView[TVIEW];
+  const tI18n = tView.data[index] as TI18n;
+  ngDevMode &&
+      assertDefined(
+          tI18n, 'Expected i18n data to be present in a given TView slot during hydration');
+
+  const firstAstNode = tI18n.ast[0];
+  if (firstAstNode) {
+    // We begin hydration at the RNode for the first child. Top-level AST nodes will always
+    // have a valid TNode, so we can use the normal `locateNextRNode` to find it. It's OK if
+    // nothing is located, as that also means that there is nothing to clean up. Downstream
+    // error handling will detect this and provide proper context.
+    const tNode = tView.data[firstAstNode.index] as TNode;
+    const rootNode = locateNextRNode(hydrationInfo, tView, lView, tNode) as Node | null;
+
+    const disconnectedNodes = initDisconnectedNodes(hydrationInfo) ?? new Set();
+    const i18nMap = hydrationInfo.i18nNodes ??= new Map<number, RNode|null>();
+    const caseQueue = hydrationInfo.data[I18N_DATA]?.[index - HEADER_OFFSET] ?? [];
+
+    prepareForHydration(
+        {hydrationInfo, i18nMap, disconnectedNodes, caseQueue},
+        {currentNode: rootNode, isConnected: true}, tI18n.ast);
+
+    // During i18n hydration, we may have discovered disconnected nodes that weren't
+    // serialized, so we need to write the expected value back.
+    hydrationInfo.disconnectedNodes = disconnectedNodes.size === 0 ? null : disconnectedNodes;
+  }
+}
+
+function prepareForHydration(
+    context: I18nHydrationContext, state: I18nHydrationState, nodeOrNodes: I18nNode|I18nNode[]) {
+  if (Array.isArray(nodeOrNodes)) {
+    for (let i = 0; i < nodeOrNodes.length; i++) {
+      prepareForHydration(context, state, nodeOrNodes[i]);
+    }
+  } else {
+    switch (nodeOrNodes.kind) {
+      case I18nNodeKind.TEXT:
+        // Claim a text node for hydration
+        markHydrationRoot(context, state, nodeOrNodes, MarkOptions.CLAIM);
```

Comment:
It's a bit hard to reason about the state (how we advance it), since the state modifications are "hidden" inside of the `markHydrationRoot` and `skipHydrationNodes` functions. I'd like to propose that we make state modification more explicit, i.e. do something like this:
```suggestion
        const currentNode = markHydrationRoot(context, state, nodeOrNodes);
        advanceCurrentNode(state, currentNode?.nextSibling);
```
Similarly, for the `skipHydrationNodes`:
```
        const nextNode = skipHydrationNodes(state, containerSize! + 1);
        advanceCurrentNode(state, nextNode);
```
In this case we can also drop `MarkOptions`.

Reply:
With this change, we can also update the names of the `prepareForHydration` (quite generic) and `markHydrationRoot` functions, for example:

- `prepareForHydration` -> `collectI18nNodesFromDom`
- `markHydrationRoot` -> `appendI18nNodeToCollection`

Reply:
In this case, the `advanceCurrentNode` isn't actually doing any advancing (since we're just passing in the node), so should it just be `setCurrentNode`?

Reply:
Yeah, `setCurrentNode` sounds great 👍